File: Sorting/quickSort/C++/quick_sort.cpp
/*Program for quick sort implementation in C++. */

#include <iostream>
#include <vector>

int partition(std::vector<int> &A, int start, int end)
{
	int temp;
	int pivot = A[end];
	int partitionIndex = start; // set partition index as start initially
	for(int i=start; i<end; i++)
	{
		if(A[i] <= pivot)
		{
			swap(&A[i], &A[partitionIndex]);
			partitionIndex++;
		}
	}
	swap(&A[partitionIndex], &A[end]);		//	partition index

	return partitionIndex;
}

void swapElements(int* elementOne, int* elementTwo)
{
	int temp = *elementOne;
	*elementOne = *elementTwo;
	*elementTwo = temp;
}
std::vector<int> quick_sort(std::vector<int> &A, int start, int end)
{
/*        std::cout<<"Current list : \n";
        for(int i=0;i<A.size();i++)
        {
                std::cout<<A[i]<<" ";
        }
        std::cout<<"\nSorted list : \n";
*/
	
	if(start<end)
	{
		int partitionIndex = partition(A,start,end);
		quick_sort(A,start,partitionIndex-1);
		quick_sort(A,partitionIndex+1, end);
	}

/*        for(int i=0;i<A.size();i++)
        {
                std::cout<<A[i]<<" ";
        }
*/        return A;
}

int main()
{
        std::cout<<"\nTest\n";
        std::vector<int> arr={10,2,23,-4,235,56,2,6,5,5,5,23,-4,346,-56,81,43,654,435,-54};
        quick_sort(arr, 0, arr.size()-1);
        std::cout<<"\n\n\n";
        for(int i=0;i<arr.size();i++)
        {
                std::cout<<arr[i]<<" ";
        }
        return 0;
}

File: Sorting/quickSort/Rust/quick-sort.rs
pub fn quick_sort<T: Ord>(arr: &mut [T]) {
    let len = arr.len();
    _quick_sort(arr, 0, (len - 1) as isize);
}

fn _quick_sort<T: Ord>(arr: &mut [T], low: isize, high: isize) {
    if low < high {
        let p = partition(arr, low, high);
        _quick_sort(arr, low, p - 1);
        _quick_sort(arr, p + 1, high);
    }
}

fn partition<T: Ord>(arr: &mut [T], low: isize, high: isize) -> isize {
    let pivot = high as usize;
    let mut store_index = low - 1;
    let mut last_index = high;

    loop {
        store_index += 1;
        while arr[store_index as usize] < arr[pivot] {
            store_index += 1;
        }
        last_index -= 1;
        while last_index >= 0 && arr[last_index as usize] > arr[pivot] {
            last_index -= 1;
        }
        if store_index >= last_index {
            break;
        } else {
            arr.swap(store_index as usize, last_index as usize);
        }
    }
    arr.swap(store_index as usize, pivot as usize);
    store_index
}
fn main() {
    println!("Sort numbers ascending");
    let mut numbers = [4, 65, 2, -31, 0, 99, 2, 83, 782, 1];
    println!("Before: {:?}", numbers);
    quick_sort(&mut numbers);
    println!("After:  {:?}\n", numbers);
}
